blocks/graph.py

```diff
@@ -354,3 +384,51 @@ def apply_noise(computation_graph, variables, level, seed=None):
         replace[variable] = (variable +
                              rng.normal(variable.shape, std=level))
     return computation_graph.replace(replace)
+
+
+def apply_dropout(computation_graph, variables, prob=0.5, rng=None, seed=None):
+    """Returns a graph to variables in a computational graph.
+
+    Parameters
+    ----------
+    computation_graph : instance of :class:`ComputationGraph`
+        The computation graph.
+    variables : list of :class:`~tensor.TensorVariable`
+        Variables to be dropped out.
+    prob : float
+        Probability of dropping out. If you want to apply the dropout
+        with different probabilities for different layers, call it
+        several times.
+    rng : :class:`~theano.sandbox.rng_mrg.MRG_RandomStreams`
+        Random number generator.
+    seed : int
+        Random seed to be used if `rng` was not specified.
+
+    Notes
+    -----
+    For more information, see [DROPOUT]_.
+
+    .. [DROPOUT] Hinton et al. *Rate MethodImproving neural networks
+    by preventing co-adaptation of feature detectors*, arXiv:1207.0580.
+
+    """
+    if not rng and not seed:
+        seed = config.default_seed
+    if not rng:
+        rng = MRG_RandomStreams(seed)
+    replacements = {var: var * rng.binomial(var.shape, p=prob,
+                                            dtype=floatX)
+                    for var in variables}
+    variables_set = set(variables)
+    weights = []
+    for var in computation_graph.variables:
+        if hasattr(var.tag, 'roles') and OUTPUT in var.tag.roles and \
+                set(var.owner.inputs).intersection(variables_set):
+            for input in var.owner.inputs:
+                if WEIGHTS in input.tag.roles:
+                    weights.append(input)
+
+    weights_replace = {weight: weight / prob for weight in weights}
```

Comment:
As written, I think this would be not only more complicated but incorrect, since at test time you want to scale by the inverse of the *inclusion* probability rather than the drop probability. It helps to think about a probability other than 0.5; if you are dropping 20% of the time then each input is appearing, on average, 80% of the time, so you want to scale it to 80% of it's actual value at test time (or 125% at training time, or whatever).

But rescaling the inputs is far less messy, I think (you still want to use `1 - prob` though).